src/async_impl/multipart.rs

```diff
@@ -0,0 +1,390 @@
+      std::mem::replace(&mut self.fields, Vec::new())
+    }
+}
+
+impl fmt::Debug for Form {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Form")
+            .field("boundary", &self.boundary)
+            .field("parts", &self.fields)
+            .finish()
+    }
+}
+
+
+/// A field in a multipart form.
+pub struct Part {
+    value: Body,
+    mime: Option<Mime>,
+    file_name: Option<Cow<'static, str>>,
+    headers: HeaderMap,
+}
+
+impl Part {
+    /// Makes a text parameter.
+    pub fn text<T>(value: T) -> Part
+    where T: Into<Cow<'static, str>>,
+    {
+        let body = match value.into() {
+            Cow::Borrowed(slice) => Body::from(slice),
+            Cow::Owned(string) => Body::from(string),
+        };
+        Part::new(body)
+    }
+
+    /// Makes a new parameter from arbitrary bytes
+    pub fn bytes<T>(value: T) -> Part
+    where T: Into<Cow<'static, [u8]>>
+    {
+        let body = match value.into() {
+            Cow::Borrowed(slice) => Body::from(slice),
+            Cow::Owned(vec) => Body::from(vec),
+        };
+        Part::new(body)
+    }
+
+    /// Makes a new parameter from an arbitrary stream.
+    pub fn stream<T: Stream + Send + 'static>(value: T) -> Part
+    where hyper::Chunk: std::convert::From<<T as futures::Stream>::Item>,
+          <T as futures::Stream>::Error: std::error::Error + Send + Sync {
+      Part::new(Body::wrap(hyper::Body::wrap_stream(value)))
+    }
+
+    fn new(value: Body) -> Part {
+        Part {
+            value: value,
+            mime: None,
+            file_name: None,
+            headers: HeaderMap::default()
+        }
+    }
+
+    /// Tries to set the mime of this part.
+    pub fn mime_str(mut self, mime: &str) -> ::Result<Part> {
+        self.mime = Some(try_!(mime.parse()));
+        Ok(self)
+    }
+
+    // Re-enable when mime 0.4 is available, with split MediaType/MediaRange.
+    #[cfg(test)]
+    fn mime(mut self, mime: Mime) -> Part {
+        self.mime = Some(mime);
+        self
+    }
+
+    /// Sets the filename, builder style.
+    pub fn file_name<T: Into<Cow<'static, str>>>(mut self, filename: T) -> Part {
+        self.file_name = Some(filename.into());
+        self
+    }
+
+    /// Returns a reference to the map with additional header fields
+    pub fn headers(&self) -> &HeaderMap {
+        &self.headers
+    }
+
+    /// Returns a reference to the map with additional header fields
+    pub fn headers_mut(&mut self) -> &mut HeaderMap {
+        &mut self.headers
+    }
+}
+
+impl fmt::Debug for Part {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Part")
+            .field("value", &self.value)
+            .field("mime", &self.mime)
+            .field("file_name", &self.file_name)
+            .field("headers", &self.headers)
+            .finish()
+    }
+}
+
+#[derive(Debug, Clone)]
+struct AttrCharEncodeSet;
```

Comment:
Since this and `PercentEncoding` methods are the same as the sync version, so that we don't have duplicate code, mind updating `reqwest::multipart` to import and use these?

Reply:
Yeah, there it'll require changing them to accept either an Enum or Trait that both the sync and async versions of Part implement. Or do you have something else in mind?

Reply:
I don't think it'd require that, since neither `AttrCharEncodeSet` nor `PercentEncode` are part of the public API...